pkgs.formats: Add HOCON format generator

h7x4 opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Taking some of the lessons learnt from #246115, I've made a generator for the [HOCON format](https://github.com/lightbend/config/blob/main/HOCON.md). This generator aims to support (almost) all the features of the HOCON format, including things like substitutions from envvars, different kinds of include statements both for declarative and mutual includes, append statements, etc. 

HOCON does not seem to have a definitive governing body, nor a complete formal specification, so I've chosen to base my implementation upon the best I could find: https://github.com/lightbend/config/blob/main/HOCON.md. This means that there's bound to be some compatibility issues with certain implementations (in fact, I've already stumbled upon one with the include types when I tried writing the validator using [the most popular go implementation](https://github.com/gurkankaymak/hocon)), but I think this is unavoidable to some degree.

Up until now, I've only found these three that are using the program, as well as a pending package. Most of them seem to have copied each others implementation, so backwards compatability can probably be combined:

- `nixos/modules/services/networking/jitsi-videobridge.nix`
- `nixos/modules/services/networking/jicofo.nix`
- `nixos/modules/services/networking/jibri/default.nix`
- `nixos/modules/services/web-apps/tachidesk-server.nix` (#257733)

## TODO:

- [x] write generator
- [x] write tests
- [x] replace existing implementations treewide
- [x] provide some sort of backwards compatibility or warning to smooth migration from the existing implementations
  (I think the only thing we'd need to handle would be `__hocon_envvar` and `__hocon_unquoted_string`)

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Some tests that now use this generator either directly or indirectly:

```bash
nix build ".#pkgs.tests.pkgs-lib.hocon.comprehensive"
nix build ".#pkgs.tests.pkgs-lib.hocon.backwards-compatibility"
nix build ".#nixosTests.jibri" # This one is failing for seemingly unrelated reasons
nix build ".#nixosTests.jitsi-meet"
nix build ".#nixosTests.prometheus-exporters.jitsi"
nix build ".#nixosTests.suwayomi-server.with-auth"
nix build ".#nixosTests.suwayomi-server.without-auth"
```

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

h7x4 committed:
formats.hocon: init

h7x4 committed:
formats.hocon: add tests

h7x4 committed:
treewide: use `formats.hocon`

roberth reviewed (approved):
This looks well thought out, follows an established implementation strategy (see libconfig), has tests, and the Nix code looks good to me.
Although I don't have the opportunity to dive into the intricacies of HOCON, to "definitively" claim that all the details are good, I am confident that this code is maintainable.

h7x4 committed:
formats.hocon: add backwards compatibility

h7x4 commented:
Okay, this should be completely ready now.

pings for maintainers of affected modules:

- `nixos/modules/services/networking/{jitsi-videobridge.nix,jicofo.nix,jibri/default.nix}`: Jitsi team - @cleeyv, @ryantm, @Lassulus, @yayayayaka
- `nixos/modules/services/web-apps/suwayomi-server.nix`: @RatCornu

h7x4 commented:
@ofborg test jibri jitsi-meet prometheus-exporters.jitsi suwayomi-server.with-auth suwayomi-server.without-auth

h7x4 committed:
CODEOWNERS: add h7x4 to pkgs.formats.hocon

infinisil commented on pkgs/pkgs-lib/formats/hocon/src/src/main.rs:
```diff
@@ -0,0 +1,234 @@
+            HOCONValue::Null => "null".to_string(),
+            HOCONValue::Bool(b) => b.to_string(),
+            HOCONValue::Int(n) => n.to_string(),
+            HOCONValue::Float(n) => format!("{:?}", n),
```
Are `Debug` implementations guaranteed to be stable in Rust? Otherwise can this be `{}` instead?

h7x4 replied:
The debug implementations for structs in `std` are unfortunately marked as unstable. The reason I'm using the debug for float specifically, is because the normal `Display` implementation for some reason drops `.0` for floats that are whole numbers. When parsed by HOCON, this would then change its type.

I was playing with the idea of copying the implementation to pin it, but it seems to depend on a tree of abstractions that is not exposed, so it feels like it costs more than the reward. Even though it's marked as unstable, I think they are unlikely to change any time soon.

infinisil replied:
According to https://github.com/lightbend/config/blob/main/HOCON.md, it seems to use the same numbers as JSON, including the fact that there's no distinction between floats and ints:

> - values have possible types: string, number, object, array, boolean, null
> - allowed number formats matches JSON; as in JSON, some possible floating-point values are not represented, such as NaN

So it seems like it should be fine to ditch `HOCONValue::{Int,Float}` in favor of `HOCONValue::Number` and use `serde_json` to serialise its value.

infinisil commented on pkgs/pkgs-lib/formats/hocon/src/src/main.rs:
```diff
@@ -0,0 +1,234 @@
+    }
+}
+
+fn escape_string(s: &str) -> String {
+    s.replace('\\', "\\\\").replace('\"', "\\\"").replace('\n', "\\n").replace('\t', "\\t")
+}
```
Would be great to have a link listing that only these characters need to be escaped

h7x4 replied:
I don't think there's a single place in the docs that points this out. I believe these were my reasons for the escapes. Should I add it as a comment?

- `\` is there for ensuring existing escapes are correct.
- `"` is there because the entire string will get wrapped.
- `\n` is there because wrapping the string with `"` implies a single line string.
- `\t` might not even be necessary. Maybe I should remove it?

infinisil replied:
Oh actually it does mention this:

> - quoted strings are in the same format as JSON strings

So here it also seems like we should be able to just use `serde_json`'s string encoding

h7x4 replied:
I've fixed it up. I discovered that it's a bit ambiguous how substitutions should be escaped, if at all, so I've left it unescaped for now. Hoping that the `checkPhase` catches any invalid configuration left.

infinisil commented on pkgs/pkgs-lib/formats/hocon/test/default.nix:
```diff
@@ -0,0 +1,5 @@
+{ pkgs, ... }:
+{
+  comprehensive = pkgs.callPackage ./comprehensive { };
+  backwards-compatibility = pkgs.callPackage ./backwards-compatibility { };
```
Here's a perhaps cleaner alternative to `_disableDeprecationWarnings`:

```suggestion
  backwards-compatibility =
    let
      pkgsNoWarn = pkgs.extend (final: prev: {
        lib = prev.lib.extend (libFinal: libPrev: {
          trivial = libPrev.trivial // {
            warn = msg: v: v;
          };
        });
      });
    in pkgsNoWarn.callPackage ./backwards-compatibility { };
```

h7x4 replied:
This doesn't work, and I'm having a hard time figuring out why. Could it be that the `lib` used for `callPackage` is already captured in a closure somewhere? (ftr, I've also tried directly overriding `lib.warn` without going through `lib.trivial`, but it yields the same result)

h7x4 replied:
Pushed the broken changes as a fixup commit

infinisil reviewed (approved):
Overall this is looking great, I'm impressed by the code quality! Just got some minor suggestions, but I'm happy to merge

infinisil committed:
pkgs-lib: Make `lib` overlays be propagated

This is useful because the tests in `pkgs-lib` can mock out certain
`lib` functions like this using a `lib` overlay.

h7x4 commented:
> Found the problem, pushed the fix :D

Cool! I've reordered and rebased the fixup to make more sense.

infinisil commented:
Looking good, let's merge!

infinisil commented:
Forgot to mention that this should be documented! See https://github.com/NixOS/nixpkgs/issues/288163
